state-chain/chains/src/sol/sol_tx_building_blocks/instruction_builder.rs

```diff
@@ -0,0 +1,275 @@
+//! This file contains a Instruction builder for the Solana chain.
+//!
+//! The builder provides a interface for users to create Raw solana
+//! Instructions and Instruction sets with some level of abstraction
+//! so the user do not need to deal with low level code in `sol_tx_building_blocks.rs`.
+
+use codec::{Decode, Encode};
+use core::str::FromStr;
+use scale_info::TypeInfo;
+use sp_core::RuntimeDebug;
+
+use sp_std::marker::PhantomData;
+
+use cf_primitives::chains::Solana;
+
+use crate::{
+	sol::{
+		api::{SolanaEnvAccountLookupKey, SolanaEnvironment},
+		compute_budget::ComputeBudgetInstruction,
+		consts::SYSTEM_PROGRAM_ID,
+		program_instructions::{ProgramInstruction, SystemProgramInstruction, VaultProgram},
+		SolAccountMeta, SolInstruction, SolPubkey,
+	},
+	DepositChannel,
+};
+
+/// Errors that can arise when building Solana Instructions.
+#[derive(Copy, Clone, Eq, PartialEq, Encode, Decode, RuntimeDebug, TypeInfo)]
+pub enum SolanaInstructionBuilderError {
+	// The current Aggkey is not set
+	CannotLookupAggKey,
+	// Cannot obtain an available Nonce Account
+	NoAvailableNonceAccount,
+	// Failed to lookup Compute Limit
+	CannotLookupComputeLimit,
+	// Failed to lookup Compute Price
+	CannotLookupComputePrice,
+	// Failed to lookup Vault Program Data Account
+	CannotLookupVaultProgramDataAccount,
+}
+
+pub struct SolanaInstructionBuilder<Environment: 'static> {
+	instructions: Vec<SolInstruction>,
+	_phantom: PhantomData<Environment>,
+}
+
+impl<Environment> Default for SolanaInstructionBuilder<Environment> {
+	fn default() -> Self {
+		Self { instructions: Default::default(), _phantom: Default::default() }
+	}
+}
+
+impl<Environment: SolanaEnvironment> SolanaInstructionBuilder<Environment> {
+	pub fn finalize(mut self) -> Result<Vec<SolInstruction>, SolanaInstructionBuilderError> {
+		let mut final_instructions = vec![
+			SystemProgramInstruction::advance_nonce_account(
+				&Environment::lookup_account(SolanaEnvAccountLookupKey::NextNonceAccount)
+					.ok_or(SolanaInstructionBuilderError::NoAvailableNonceAccount)?
+					.into(),
+				&Environment::lookup_account(SolanaEnvAccountLookupKey::AggKey)
+					.ok_or(SolanaInstructionBuilderError::CannotLookupAggKey)?
+					.into(),
+			),
+			ComputeBudgetInstruction::set_compute_unit_price(
+				Environment::compute_price()
+					.ok_or(SolanaInstructionBuilderError::CannotLookupComputePrice)?,
+			),
+			ComputeBudgetInstruction::set_compute_unit_limit(
+				Environment::compute_limit()
+					.ok_or(SolanaInstructionBuilderError::CannotLookupComputeLimit)?,
+			),
+		];
+
+		final_instructions.append(&mut self.instructions);
+
+		Ok(final_instructions)
+	}
+
+	pub fn fetch_from(
```

Comment:
I think it would be better if this is done inside the builder. 
We can do a match on the deposit_channel.asset, and decide the correct Instruction we want to use.
This way we will have 1 fetch_from call and we can mix and match both native and Token

Reply:
right, that probably sounds better 👍